docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,174 @@
+// a node selector requirement.
+type NodeSelectorOperator string
+
+const (
+	NodeSelectorOpIn           NodeSelectorOperator = "In"
+	NodeSelectorOpNotIn        NodeSelectorOperator = "NotIn"
+	NodeSelectorOpExists       NodeSelectorOperator = "Exists"
+	NodeSelectorOpDoesNotExist NodeSelectorOperator = "DoesNotExist"
+	NodeSelectorOpGt           NodeSelectorOperator = "Gt"
+	NodeSelectorOpLt           NodeSelectorOperator = "Lt"
+)
+```
+
+### Changes to PodSpec
+
+We will add two fields to `PodSpec`:
+
+```go
+// NodeConstraints specifies hard scheduling requirements for the Pod.
+NodeConstraints *NodeSelector `json:nodeConstraints,omitempty"`
+// NodePreferences specifies which nodes are preferred among those that meet all
+// of the hard scheduling requirements (resource request, constraints, etc.).
+// We don't use *NodeSelector as the type here, so as to make it clear that
+// we are not producing a single set of nodes as the result, but rather describing
+// several sets of nodes. A node that is in more such sets is
+// considered to be "more preferred" than a node that is in fewer.
+NodePreferences []NodeSelectorTerm `json:nodePreferences,omitempty"`
+```
+
+Unfortunately, the name of the existing `map[string]string` field in PodSpec is `NodeSelector`
+and we can't change it since this name is part of the API. Hopefully this won't
+cause too much confusion.
+
+## Backward compatibility
+
+When we add the `NodeConstraints` and `NodePreferences` fields to PodSpec,
+we will deprecate, but not remove, the current field in PodSpec
+
+```go
+`NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+```
+
+Old version of the scheduler will ignore the new `nodeConstraints` and `nodePreferences` fields.
+New versions of the scheduler will apply their scheduling predicates to both `nodeConstraints` and `nodeSelector`,
+i.e. the pod can only schedule onto nodes that are in both sets. We will not
+attempt to convert between `nodeConstraints` and `nodeSelector`.
+
+Old versions of non-scheduling clients will not know how to do anything semantically meaningful
+with `nodeConstraints` and `nodePreferences`, but we don't expect that this will cause a problem.
+
+See [this comment](https://github.com/kubernetes/kubernetes/issues/341#issuecomment-140809259)
+for more discussion.
+
+## Implementation plan
+
+1. Add the `NodeSelector`, `NodeConstraints`, and `NodePreferences` types to the API
+1. Add the `NodeConstraints` and `NodePreferences` fields to `PodSpec`
+1. Implement scheduler predicate that takes `NodeConstraints` into account
+1. Implement scheduler priority function that takes `NodePreferences` into account
+1. Mark `PodSpec.NodeSelector` field as deprecated
+
+## Future work
+
+Are there any other fields we should convert from `map[string]string` to `NodeSelector`?
+
+## Related issues
+
+The main related issue is #341. Issue #367 is also related. Those issues reference other
```

Comment:
I didn't bother to do links in these proposals. If someone wants to send a followup PR to fix it they're welcome to.